components/console/test_apps/console/main/test_console.c

```diff
@@ -102,3 +107,52 @@ TEST_CASE("esp console init/deinit test, minimal config", "[console]")
     TEST_ESP_OK(esp_console_cmd_register(&cmd));
     TEST_ESP_OK(esp_console_deinit());
 }
+
+// Enter "quit" to exit REPL environment
+/* Marked as ignore since it cannot run as a normal unity test case
+   ran separately in test_console_repl  */
+TEST_CASE("esp console repl test with context", "[console][ignore]")
+{
+    esp_console_repl_config_t repl_config = ESP_CONSOLE_REPL_CONFIG_DEFAULT();
+    esp_console_dev_uart_config_t uart_config = ESP_CONSOLE_DEV_UART_CONFIG_DEFAULT();
+    TEST_ESP_OK(esp_console_new_repl_uart(&uart_config, &repl_config, &s_repl));
+
+    const esp_console_cmd_t cmds[] = {
+        {
+            .command = "hello",
+            .help = "Print Hello World",
+            .hint = NULL,
+            .func = do_hello_cmd,
+        },
+        {
+            .command = "hello-c0",
+            .help = "Print Hello World in NULL context",
+            .hint = NULL,
+            .func_context = do_hello_cmd_with_context,
+        },
+        {
+            .command = "hello-c1",
+            .help = "Print Hello World in context c1",
+            .hint = NULL,
+            .func_context = do_hello_cmd_with_context,
+        },
+        {
+            .command = "hello-c2",
+            .help = "Print Hello World in context c2",
+            .hint = NULL,
+            .func_context = do_hello_cmd_with_context,
+        },
+    };
+    for (int i=0; i < sizeof(cmds) / sizeof(cmds[0]); i++) {
+        TEST_ESP_OK(esp_console_cmd_register(&cmds[i]));
+    }
+    TEST_ESP_OK(esp_console_cmd_set_context(cmds[2].command, cmds[2].command));
+    TEST_ESP_OK(esp_console_cmd_set_context(cmds[3].command, cmds[3].command));
+    TEST_ASSERT_EQUAL(esp_console_cmd_set_context(NULL, NULL), ESP_ERR_INVALID_ARG);
+    TEST_ASSERT_EQUAL(esp_console_cmd_set_context("invalid", NULL), ESP_ERR_NOT_FOUND);
+    TEST_ESP_OK(esp_console_cmd_register(&s_quit_cmd));
+    TEST_ESP_OK(esp_console_register_help_command());
+
+    TEST_ESP_OK(esp_console_start_repl(s_repl));
```

Comment:
I think for unit test purposes, you don't need to start the whole REPL. You can call `esp_console_run` and verify that the expected callback got called with the expected context. (Either by setting some variable from the callback, or by temporarily substituting `stdout` using `open_memstream` and then asserting that the expected string was printed by the callback.)

Reply:
Done, I think.